Fetch valid RSEs and keep it up-to-date in MSPileup

amaltaro opened the pull request:
Fixes #11498 (sort of!)

#### Status
In development

#### Description
This PR provides the following:
* reads a configurable rudio Disk expression `rucioDiskExpression` to be used to find valid and allowed Rucio RSEs from Rucio (which relies on a memory cache set to expiry every 12h, by default)
* it enforces RSE validation when creating a new pileup document
* when user updates a pileup object, validation of the parameters will be performed as well
* when MSPileupTasks updates a pileup object, validation of the document is skipped (given that it's already loaded from the database)

#### Is it backward compatible (if not, which system it affects?)
YES

#### Related PRs
None

#### External dependencies / deployment changes
None

amaltaro committed:
MSPileup: validate list of RSEs in both create/update calls

Protect validRSEs variable by threading lock

Using lock as a context manager

alternative to fetch the list of RSEs for each create/update call

Fetch list of RSEs for each create request

Validate user-input data for updatePileup call

Provide flag to enable or not updatePileup

rseList has to be optional for updatePileup

change default value for validate to True

amaltaro commented:
@vkuznet Valentin, could you please have a look at this initial implementation for the RSE validation? If it's looking Okay for you, I can update unit tests as well.

vkuznet reviewed (changes requested):
Alan, I think you incorrectly provide the logic to MSPileup, while in my opinion it should go to MSPIleupTaskManager (i.e. the update action should be done by task manager rather HTTP server). This is HTTP service, as such it only provides HTTP methods like GET/POST/DELETE/PUT. From your your changes I see multiple flaws:
- you changed the behavior of pileup thread which should only retrieve the status
- unless you want to provide HTTP way to modify validRSEs as an external request this implementation does not belong to MSPileup and should go to MSPileupTaskManager and executed as independent task
- MSPileup class is rucio free since it does not involved in any data placement operation

amaltaro committed:
Only add data to the cache in Rucio wrapper when requested for

amaltaro commented:
@vkuznet Valentin, as we discussed over Zoom, RSEs need to be validated between the HTTP request and the action to persist it in the database, so this MSPileup service is the proper place to have the definition of the valid RSEs on (instead of the Tasks instance).

I updated this PR as follows:
* 2nd commit provides a fix on how we actually use MemoryCache in the Rucio wrapper and ensuring it won't grow when we are not using the cache
* 3rd commit provides a lock to the MSPileupData instance variable `validRSEs`, to avoid possible race conditions between reading and updating the list

The standard libraries don't seem to provide any way to perform separate read/write locks (blocking reads only based on the write lock). For that, we would have to implement a custom class, [example](https://gist.github.com/tylerneylon/a7ff6017b7a1f9a506cf75aa23eacfd6) (but I didn't check the full implementation).

I am still not very happy with the use of this lock and maybe the best would be to fetch RSEs from the Rucio wrapper cache for every createPileup/updatePileup call and pass that over for the validation.

amaltaro commented:
And with the 4th commit - which is very much incomplete - I show how we could get rid of the threading lock mechanism. In plain english, using the Rucio wrapper cache, we could fetch the list of RSEs whenever there is an incoming Create or Update pileup object call.

Note the placeholder for the updatePileup method, which turns out not to enforce any user data validation at the current moment.

vkuznet commented:
> And with the 4th commit - which is very much incomplete - I show how we could get rid of the threading lock mechanism. In plain english, using the Rucio wrapper cache, we could fetch the list of RSEs whenever there is an incoming Create or Update pileup object call.
> 
> Note the placeholder for the updatePileup method, which turns out not to enforce any user data validation at the current moment.

So, you are gradually converging to what I suggested, having rucio call for every HTTP request and passing around rses. Yes, in this implementation we do not need locks and it will be free from racing conditions.

Meanwhile you'll do your changes, please separate `updateRSE()` and `status()` calls. The former updates RSE, the latter get the status, therefore the former should not include the latter. I suggest to remove status call from updateRSE and use them separately in pileup function in MSManager.

amaltaro commented:
> So, you are gradually converging to what I suggested, having rucio call for every HTTP request and passing around rses. Yes, in this implementation we do not need locks and it will be free from racing conditions.

Okay, I take it that the 4th commit provides an approach that you like the better as well. 
I will modify the code appropriately, but it might be that I won't get it ready by the end of the day tomorrow. Thanks Valentin.

amaltaro commented:
@vkuznet Valentin, I still have to check the jenkins report and fix the unit tests. But if you have any important feedback on the latest implementation, please let me know.

amaltaro committed:
fix unit tests

Separate and fix MSPileup unit tests

fix testMSPileupTasksWithMockApi

unit test function to create doc

amaltaro commented:
@vkuznet Valentin, there are a couple of MSPileupTasks unit tests that now fail and the reason is that now document update is validated as well, while the current master code does not run any validation on top of the user-provided data (which in our case could be coming from MSPileupTasks as well).

So I would like to ask your opinion on the following potential implementations and/or behaviors:
* first question is whether we should always validate data being updated to the documents. My opinion is that end-users must have everything updated; while data provided by the microservice itself don't need to go through the whole MSPileupObj validation
* it looks like we read the document from MongoDB twice as well (one too many!). First when executing the MSPileupTasks tasks; second when calling MSPileupData.updatePileup.
* to avoid the double read and thorough document validation for microservice-originated data, we could create a "private" method in MSPileup that will be used only by MSPileupTasks.

Please let me know what your thoughts are. Others are always welcomed to chime in as well, of course!

vkuznet commented:
In my opinion, the validation should be done in all steps, regardless if user create or update the document. Lack of validation in update step may lead to silly mistakes from end-user, e.g. a typo, which will provide invalid docs at the end.

For other two items you asked for I need more context. But I think you are talking about the following block https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/MicroService/MSPileup/MSPileupData.py#L130-L143 If so, I think it would be much cleaner to remove it to avoid double read and let code fail in update if document does not exist (if you'll modify the code you may change `insert` to `update` in failed message on line 151.

amaltaro commented:
> In my opinion, the validation should be done in all steps, regardless if user create or update the document. Lack of validation in update step may lead to silly mistakes from end-user, e.g. a typo, which will provide invalid docs at the end.

Yes, we agree on that!

However, there are two scenarios where a pileup document can be updated:
1) from user-provided arguments, through the RESTful API.
2) from microservice-provided arguments, in this case MSPileupTasks, which loads every single document from the database, change it if needed and persist the up-to-date document back in the database.

So, my question is specific to the case 2). After MSPileupTasks performed all the actions on a given MSPileupObj object and it's about to make a call to `updatePileup` to get the updated document persisted in the database, do you think we should pass the new dictionary through the MSPileupObj.validate call, such that we verify the data before persisting it? Or would you say that we should be safe (enough), assuming that the microservice is doing the right thing, thus simply by-passing any further MSPileupObj.validate call?

If you think we should not validate it, then I believe we would need to create a "private" method, e.g. `_updatePileup`, which will be used only by MSPileupTasks, such that the data structure passed to be persisted isn't validated against the MSPileupObj class.

If I was not clear enough, maybe a 10min Zoom chat will be better.

vkuznet commented:
Alan, I still do not understand why do you need any private method, why you can't just add additional parameter to updatePileup to skip validation and make it obligatory by default, e.g. `def updatePileup(self, pdict, validate=True)`. This way there is no need for another private method, and you can clearly distinguish that all HTTP clients will by default use validation and MSPileupTasks can provide this extra parameter to skip it.

amaltaro committed:
Mock Rucio in MSCore module

amaltaro commented:
I think I am done with changes to this PR and it's ready for a full review.

vkuznet commented on src/python/WMCore/MicroService/MSPileup/MSPileup.py:
```diff
@@ -80,28 +80,30 @@ def queryDatabase(self, query, projection=None):
         """
         self.authMgr.authorizeApiAccess('ms-pileup', 'update')
-        return self.mgr.updatePileup(pdict)
+        rseNames = self.rucio.evaluateRSEExpression(self.diskRSEExpr, useCache=True)
```
Alan, I thought we agree to not use the cache, while here it is the opposite. Could you please elaborate on this `useCache=True` option.

amaltaro replied:
No, we actually agreed to use the Rucio wrapper cache, specially because this call is tied to EVERY HTTP call, so we better use it not to increase latency in the HTTP calls (also because the RSE expression is always the same!).

vkuznet replied:
The discussion was about effect of the cache to HTTP service which should be simple. Said that, I will not argue here until we know that cache really significantly contribute to RAM usage. Let's proceed with `usageCache=True` and keep eye on RAM usage.

amaltaro replied:
I don't think we need to be concerned with this because cache will be a dictionary with a single key/value pair. Any updates to the cache will be void because the key will already exist.

vkuznet commented on src/python/WMCore/MicroService/MSPileup/MSPileupData.py:
```diff
@@ -113,37 +114,58 @@ def createPileup(self, pdict):
         return []
 
-    def updatePileup(self, doc):
+    def updatePileup(self, doc, rseList=None, validate=False):
```
I thought that default should always be `validate=True` and only upstream code will explicitly change this parameter.

amaltaro replied:
Just for convenience, given that with this default, there is only 1 place calling it that needs to be changed (MSPileup). While the inverted logic would require changes in a handful of calls from MSPileupTasks (and maybe other places). Other than that, I have no strong preference and I could revert the logic if you want. Just let me know.

vkuznet replied:
Regardless how much changes should be done, the default should be right here. I strongly suggest to make `validate=True` as default and make explicit calls when this should be skipped.

vkuznet commented on src/python/WMQuality/Emulators/EmulatedUnitTestCase.py:
```diff
@@ -45,7 +45,7 @@ def setUp(self):
         TODO: parameters to turn off emulators individually
         """
-
+        print(f"Running setUp with emulated services. ")
```
why print and not logger call?

amaltaro replied:
No strong preference as well, but for many tests the logger buffer gets many many records, so a line like that is harder to see than if we get it in the stdout. I am fine to change it though, if you prefer.

vkuznet replied:
The strange thing that you always made similar argument when I used print in a code. I followed your suggestion here :) since if codebase use logger everywhere it seems logical to use it here too.

amaltaro replied:
I think for unit tests it gives us a good separation from the actual WMCore code, thus stdout would have only unit test related log records.

vkuznet commented on test/data/Mock/RucioMockData.json:
```diff
@@ -1404,6 +1404,7 @@
   "account": "ms-pileup"
  }],
  "evaluateRSEExpression": ["T2_XX_CERN"],
+ "evaluateRSEExpression:[('useCache', True)]": ["rse1", "rse2"],
```
should we use useCache=False instead?

amaltaro replied:
We should use whatever signature of the call we have in place. Given that we are using the cache, it has to be True here as well.

vkuznet replied:
fine.

amaltaro commented:
@vkuznet I think I covered most of your concerns, please have another look at your convenience.

vkuznet reviewed (approved):
Alan, changes look good to me, please proceed with merge.

amaltaro commented:
Thank you very much for the review over the weekend, Valentin.

amaltaro commented:
Tag 2.2.0rc1 should have these changes, plus MSAuth, plus fix for monitor.sh. Build pipeline is still running:
https://github.com/dmwm/WMCore/actions/runs/4398024756

Once we confirm CRUD operations are functional, I think we can then have another iteration of P&R tests.
